### PR TITLE
#914 Implemented a test run depending on network type

### PR DESCRIPTION
**Pull request by artemjackson**
This change is [Reviewable](https://reviewable.io/reviews/thaliproject/thali_cordovaplugin/922)

### CODE REVIEW

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 1. task. (30fac5a)

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (30fac5a)

See https://github.com/ThaliTester/TestResults/tree/8233723530fac5a__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

File: test/www/jxcore/bv_tests/testThaliMobileNative.js [hunk 1]
```diff
@@ -1,5 +1,14 @@
 'use strict';
 
+// Issue #419
+var ThaliMobile = require('thali/NextGeneration/thaliMoblie');
+if (global.NETWORK_TYPE === ThaliMobile.networkTypes.NATIVE ||
+    global.NETWORK_TYPE === ThaliMobile.networkTypes.BOTH) {
+  // keep going
+} else {
```
Comment (andrew-aladev) [inline test/www/jxcore/bv_tests/testThaliMobileNative.js]: Will this switch be extended with `else if` in future? We can use `if ( a !== b ) { return }` if not.

**Comment by ThaliTester** [comment]
Test 8233723530fac5a(30fac5a) has failed

See https://github.com/ThaliTester/TestResults/tree/8233723530fac5a__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the fail logs

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 1. task. (d68dad2)

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 2. task. (c8e499b)

File: test/www/jxcore/bv_tests/testThaliMobileNative.js (hunk 1 above)
Comment (artemjackson) [inline test/www/jxcore/bv_tests/testThaliMobileNative.js]: A good point

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 3. task. (c72b2f9)

**Comment by andrew-aladev** [comment]
:lgtm:

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-:-KPr7bjnBKYpMGWRHyDk:bnfp4nl)*

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (d68dad2)

See https://github.com/ThaliTester/TestResults/tree/82337235d68dad2__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
Test 82337235d68dad2(d68dad2) has failed

See https://github.com/ThaliTester/TestResults/tree/82337235d68dad2__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the fail logs

**Comment by ThaliTester** [comment]
Test 82337235c8e499b(c8e499b) has failed

See https://github.com/ThaliTester/TestResults/tree/82337235c8e499b__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the fail logs

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (c72b2f9)

See https://github.com/ThaliTester/TestResults/tree/82337235c72b2f9__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
Test 82337235c72b2f9(c72b2f9) has failed

See https://github.com/ThaliTester/TestResults/tree/82337235c72b2f9__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the fail logs

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 1. task. (c858ce8)

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (c858ce8)

See https://github.com/ThaliTester/TestResults/tree/82337235c858ce8__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
Test 82337235c858ce8(c858ce8) has failed

See https://github.com/ThaliTester/TestResults/tree/82337235c858ce8__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the fail logs

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 2. task. (741286d)

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (741286d)

See https://github.com/ThaliTester/TestResults/tree/82337235741286d__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
Test 82337235741286d(741286d) has failed

See https://github.com/ThaliTester/TestResults/tree/82337235741286d__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the fail logs

**Comment by artemjackson** [comment]
Hmmm, I saw this error while submitting #913 but it suddenly dissapeared

**Comment by yaronyg** [comment]
Once the issues I identified below are resolved and this passes CI (transient bugs not withstanding) then :lgtm: 

---

Reviewed 1 of 6 files at r1, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/www/jxcore/UnitTest_app.js, line 55 \[r3\]](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-KPtMnePsSzkCPS1THFv:-KPtMnePsSzkCPS1THFw:b1mqsj0) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/741286dd0fec2d05f08b0a89ae7d4a07c0fd5c95/test/www/jxcore/UnitTest_app.js#L55)):*
> ```JavaScript
>   global.Mobile = require('./lib/wifiBasedNativeMock.js')();
>   networkTypes = [
>     ThaliMobile.networkTypes.NATIVE,
> ```

Please add in ThaliMobile.networkType.WIFI to this list. I  had updated the description of #914 a few days ago to explain why it can now be safely added. But I guess that updating the description doesn't send out mail so in the future I'll notify you if I make such a change. Sorry.

---

*[test/www/jxcore/bv_tests/testThaliMobileNative.js, line 5 \[r3\]](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-KPtNWLy8D6ANH7DMLR2:-KPtNWLzzeHFyrfgIODF:b-gaqbdp) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/741286dd0fec2d05f08b0a89ae7d4a07c0fd5c95/test/www/jxcore/bv_tests/testThaliMobileNative.js#L5)):*
> ```JavaScript
> // Issue #419
> var ThaliMobile = require('thali/NextGeneration/thaliMobile');
> if (global.NETWORK_TYPE !== ThaliMobile.networkTypes.NATIVE &&
> ```

Why not just check if this is WiFi and then return? Why do the negative checks? Same with the checks below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-:-KPtOBpfgdVlGj1HxGw1:b77f3ts)*

**Comment by artemjackson** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/www/jxcore/bv_tests/testThaliMobileNative.js, line 5 \[r3\]](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-KPtNWLy8D6ANH7DMLR2:-KPvmpUwumtPj2Z0FZ55:bgzy2dx) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/741286dd0fec2d05f08b0a89ae7d4a07c0fd5c95/test/www/jxcore/bv_tests/testThaliMobileNative.js#L5)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> Why not just check if this is WiFi and then return? Why do the negative checks? Same with the checks below.
</details>
Currently we know that there are three network types: `WIFI`, `NATIVE` and `BOTH`.
And yes, if we want just to allow `NATIVE` and `BOTH` we can simple disallow `WIFI` like that
```
if (global.NETWORK_TYPE === ThaliMobile.networkTypes.WIFI) {
  return;
} 
```
But, if someone in future will add new networkType this new networkType will pass this check as well.
Such behaviour is not desirable. So the check you are pointed is made for future maintainability.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922)*

**Comment by artemjackson** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[test/www/jxcore/UnitTest_app.js, line 55 \[r3\]](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-KPtMnePsSzkCPS1THFv:-KPwQUiHJmhrPzv2jdvl:b-ei99eg) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/741286dd0fec2d05f08b0a89ae7d4a07c0fd5c95/test/www/jxcore/UnitTest_app.js#L55)):*
<details><summary><i>Previously, yaronyg (Yaron Y Goland) wrote…</i></summary>
> Please add in ThaliMobile.networkType.WIFI to this list. I  had updated the description of #914 a few days ago to explain why it can now be safely added. But I guess that updating the description doesn't send out mail so in the future I'll notify you if I make such a change. Sorry.
</details>
So probably It would be better to merge `if` and `else` statements content

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922)*

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 2. task. (c6facd5)

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (c6facd5)

See https://github.com/ThaliTester/TestResults/tree/82337235c6facd5__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
Test 82337235c6facd5(c6facd5) has successfully finished without an error

See https://github.com/ThaliTester/TestResults/tree/82337235c6facd5__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by yaronyg** [comment]
Reviewed 1 of 5 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[test/www/jxcore/UnitTest_app.js, line 55 \[r3\]](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-KPtMnePsSzkCPS1THFv:-KPxz5FPZW005XZV53Nk:b-p3tunw) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/741286dd0fec2d05f08b0a89ae7d4a07c0fd5c95/test/www/jxcore/UnitTest_app.js#L55)):*
<details><summary><i>Previously, artemjackson (Artem Jackson) wrote…</i></summary>
> So probably It would be better to merge `if` and `else` statements content
</details>
For now yes. Eventually it will get more complex because we will have some iOS devices that are on the guest network and so can't pass WIFI and some that are on the open network and can pass WIFI. But we'll cross that bridge when we get to it.

---

*[test/www/jxcore/bv_tests/testThaliMobileNative.js, line 5 \[r3\]](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922#-KPtNWLy8D6ANH7DMLR2:-KPxydXqCAiywWu_TrF7:bbtu0e0) ([raw file](https://github.com/thaliproject/thali_cordovaplugin/blob/741286dd0fec2d05f08b0a89ae7d4a07c0fd5c95/test/www/jxcore/bv_tests/testThaliMobileNative.js#L5)):*
<details><summary><i>Previously, artemjackson (Artem Jackson) wrote…</i></summary>
> Currently we know that there are three network types: `WIFI`, `NATIVE` and `BOTH`.
> And yes, if we want just to allow `NATIVE` and `BOTH` we can simple disallow `WIFI` like that
> ```
> if (global.NETWORK_TYPE === ThaliMobile.networkTypes.WIFI) {
>   return;
> } 
> ```
> But, if someone in future will add new networkType this new networkType will pass this check as well.
> Such behaviour is not desirable. So the check you are pointed is made for future maintainability.
</details>
This assumes that the future new value doesn't also need to be added to the banned list. No matter how you do this (letting in WiFi or banning Native/Both) I'll come up with a change that breaks you. So why don't we just stick with something simple and obvious and test for WiFi?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/thaliproject/thali_cordovaplugin/922)*

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 2. task. (2b31b5a)

**Comment by ThaliTester** [comment]
Test (Success) 82337235 build is completed (2b31b5a)

See https://github.com/ThaliTester/TestResults/tree/823372352b31b5a__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
Test 823372352b31b5a(2b31b5a) has successfully finished without an error

See https://github.com/ThaliTester/TestResults/tree/823372352b31b5a__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Comment by ThaliTester** [comment]
PR is added to the queue for testing as 6. task. (c21238e)

**Comment by ThaliTester** [comment]
Test (Fail) 82337235 build is completed (c21238e)

See https://github.com/ThaliTester/TestResults/tree/82337235c21238e__914_Implemented_a_test_run_depending_on_network_type_artemjackson/ for the logs

**Pull request closed without merging**